src/builtins/filters/number.rs

```diff
@@ -74,6 +74,105 @@ pub fn filesizeformat(value: &Value, _: &HashMap<String, Value>) -> Result<Value
+/// Additionally, the `#` modifier can be passed to some formatters as well:
+///     * `:#X` Upper hex (`42` => `0x2A`)
+///     * `:#x` Lower hex (`42` => `0x2a`)
+///     * `:#o` Octal (`42` => `0o52`)
+///     * `:#b` Binary (`42` => `0b101010`)
+pub fn format(value: &Value, args: &HashMap<String, Value>) -> Result<Value> {
+    let fmt = if let Some(fmt) = args.get("fmt") {
+        try_get_value!("format", "fmt", String, fmt)
+    } else {
+        return Err(Error::msg("Filter `format` expected an arg called `fmt`"));
+    };
+    let mut chars = fmt.chars();
+
+    if !matches!(chars.next(), Some(':')) {
+        return Err(Error::msg("Format specifiers for the `format` filter must start with `:`"));
+    }
+
+    let mut spec = chars.next().ok_or_else(|| {
+        Error::msg("Format specifiers for the `format` filter must have more than one character")
+    })?;
+
+    let alternative = if spec == '#' {
+        spec = chars.next().ok_or_else(|| {
+            Error::msg("Format strings for the `format` filter with a modifier must have a format specifier")
+        })?;
+        true
+    } else {
+        false
+    };
+
+    macro_rules! unwrap_integers {
+        ($val:expr, if $alt:ident { $alt_fmt:expr } else { $fmt:expr }, $err:expr) => {
+            if let Some(uint) = $val.as_u64() {
+                if $alt {
+                    format!($alt_fmt, uint)
+                } else {
+                    format!($fmt, uint)
+                }
+            } else if let Some(int) = $val.as_i64() {
+                if $alt {
+                    format!($alt_fmt, int)
+                } else {
+                    format!($fmt, int)
+                }
+            } else {
+                return Err($err);
+            }
+        };
+    }
+
+    let value = match spec {
+        'X' => unwrap_integers!(
+            value,
+            if alternative { "{:#X}" } else { "{:X}" },
+            Error::msg("`:X` only takes integer values")
+        ),
+        'x' => unwrap_integers!(
+            value,
+            if alternative { "{:#x}" } else { "{:x}" },
+            Error::msg("`:x` only takes integer values")
+        ),
+        'o' => unwrap_integers!(
+            value,
+            if alternative { "{:#o}" } else { "{:o}" },
+            Error::msg("`:o` only takes integer values")
+        ),
+        'b' => unwrap_integers!(
+            value,
+            if alternative { "{:#b}" } else { "{:b}" },
+            Error::msg("`:b` only takes integer values")
+        ),
+
+        'E' => {
+            let float = try_get_value!("format", "value", f64, value);
+            format!("{:E}", float)
+        }
+        'e' => {
+            let float = try_get_value!("format", "value", f64, value);
+            format!("{:e}", float)
+        }
+
+        unrecognized => {
+            return Err(Error::msg(format!("Unrecognized format specifier: `:{}`", unrecognized)))
+        }
+    };
+
+    Ok(Value::String(value))
+}
```

Comment:
That wasn't based on Jinja, it was just based on Rust, I have no knowledge of what Jinja does

Reply:
Tera is trying to roughly match what Jinja is doing. The format filter of Jinja is essentially printf (https://jinja.palletsprojects.com/en/2.11.x/templates/#format) so it would need match functionality if it has the same name. We could rename it to avoid confusion

Reply:
Implementing `printf`(https://docs.python.org/3/library/stdtypes.html#printf-style-string-formatting) is quite involved. I wonder how much of would actually get used and if a Rust programmer would know what to do with it. I can see the point of staying close to Jinja2, but matching `printf` is more like staying close to Python.

Reply:
Yep that's fine, just need to rename it so people using both are not confused by having the same name but different functionalities.